File: backend/internal/services/posts/post_usecase.go
package posts

import (
	"context"
	"fmt"
	"time"
	"y_net/internal/services/shared"

	"github.com/google/uuid"
)

type IPostUsecase interface {
	Create(ctx context.Context, post shared.Post) (uuid.UUID, error)
	GetPosts(ctx context.Context, limit int, lastCreatedAt time.Time, lastId uuid.UUID) ([]shared.Post, error)
	GetPost(ctx context.Context, id uuid.UUID) (shared.Post, error)
	Update(ctx context.Context, post shared.Post, id uuid.UUID) error
	Delete(ctx context.Context, id uuid.UUID) error
	Like(ctx context.Context, userId uuid.UUID, postId uuid.UUID) error
	Unlike(ctx context.Context, userId uuid.UUID, postId uuid.UUID) error
	UserLikedPost(ctx context.Context, userId uuid.UUID, postId uuid.UUID) (bool, error)
	GetLikes(ctx context.Context, id uuid.UUID) ([]shared.User, error)
}

type postUsecaseImpl struct {
	usecase    IPostUsecase
	repository iPostRepository
}

func NewPostUsecase() IPostUsecase {
	return &postUsecaseImpl{
		usecase:    &postUsecaseImpl{},
		repository: &postRepositoryImpl{},
	}
}

func (u *postUsecaseImpl) Create(ctx context.Context, post shared.Post) (uuid.UUID, error) {
	if (post.User == &shared.User{}) {
		return uuid.Nil, fmt.Errorf("user must not be empty")
	}
	if post.Image == "" {
		return uuid.Nil, fmt.Errorf("post image must not be empty")
	}

	id, err := u.repository.create(ctx, post)
	if err != nil {
		return uuid.Nil, err
	}

	return id, nil
}

func (u *postUsecaseImpl) GetPosts(ctx context.Context, limit int, lastCreatedAt time.Time, lastId uuid.UUID) ([]shared.Post, error) {
	posts, err := u.repository.getPosts(ctx, limit, lastCreatedAt, lastId)
	if err != nil {
		return nil, err
	}

	return posts, nil
}

func (u *postUsecaseImpl) GetPost(ctx context.Context, id uuid.UUID) (shared.Post, error) {
	post, err := u.repository.getPost(ctx, id)
	if err != nil {
		return shared.Post{}, err
	}

	return post, nil
}

func (u *postUsecaseImpl) Update(ctx context.Context, post shared.Post, id uuid.UUID) error {
	if post.Image == "" {
		return fmt.Errorf("post image must not be empty")
	}

	err := u.repository.update(ctx, post, id)
	if err != nil {
		return err
	}

	return nil
}

func (u *postUsecaseImpl) Delete(ctx context.Context, id uuid.UUID) error {
	err := u.repository.delete(ctx, id)
	if err != nil {
		return err
	}

	return nil
}

func (i *postUsecaseImpl) Like(ctx context.Context, userId uuid.UUID, postId uuid.UUID) error {
	err := i.repository.like(ctx, userId, postId)
	if err != nil {
		return err
	}

	return nil
}

func (i *postUsecaseImpl) Unlike(ctx context.Context, userId uuid.UUID, postId uuid.UUID) error {
	err := i.repository.unlike(ctx, userId, postId)
	if err != nil {
		return err
	}

	return nil
}

func (i *postUsecaseImpl) UserLikedPost(ctx context.Context, userId uuid.UUID, postId uuid.UUID) (bool, error) {
	isLiked, err := i.repository.userLikedPost(ctx, userId, postId)
	if err != nil {
		return false, err
	}

	return isLiked, nil
}

func (i *postUsecaseImpl) GetLikes(ctx context.Context, id uuid.UUID) ([]shared.User, error) {
	users, err := i.repository.getLikes(ctx, id)
	if err != nil {
		return nil, err
	}

	return users, nil
}

File: frontend/src/features/posts/domain/usecases/PostUsecase.ts
import { Post } from '@/src/features/shared/data/models/Post';
import { User } from '@/src/features/shared/data/models/User';

import { IPostRepository } from '../repositories/PostRepository';

interface IPostUsecase {
  createPost(post: Omit<Post, 'id'>): Promise<Post>;
  listPosts(limit: number, cursor?: string): Promise<Post[]>;
  getPost(id: string): Promise<Post>;
  updatePost(post: Post): Promise<boolean>;
  deletePost(id: string): Promise<boolean>;
  likePost(userId: string, postId: string): Promise<boolean>;
  unlikePost(userId: string, postId: string): Promise<boolean>;
  checkIfUserLikedPost(userId: string, postId: string): Promise<boolean>;
  getLikes(postId: string): Promise<User[]>;
}

export class PostUsecaseImpl implements IPostUsecase {
  private repository: IPostRepository;

  constructor(repository: IPostRepository) {
    this.repository = repository;
  }

  async createPost(post: Omit<Post, 'id'>): Promise<Post> {
    if (!post.image) {
      throw new Error('post image is required');
    }

    const createdPost: Post = await this.repository.createPost(post);

    return createdPost;
  }

  async listPosts(limit: number, cursor?: string): Promise<Post[]> {
    const posts: Post[] = await this.repository.listPosts(limit, cursor);

    return posts;
  }

  async getPost(id: string): Promise<Post> {
    const post: Post = await this.repository.getPost(id);

    return post;
  }

  async updatePost(post: Post): Promise<boolean> {
    if (!post.image) {
      throw new Error('post image is required');
    }

    await this.repository.updatePost(post);

    return true;
  }

  async deletePost(id: string): Promise<boolean> {
    await this.repository.deletePost(id);

    return true;
  }

  async likePost(userId: string, postId: string): Promise<boolean> {
    const didLike: boolean = await this.repository.likePost(userId, postId);

    return didLike;
  }

  async unlikePost(userId: string, postId: string): Promise<boolean> {
    const didUnlike: boolean = await this.repository.unlikePost(userId, postId);

    return didUnlike;
  }

  async checkIfUserLikedPost(userId: string, postId: string): Promise<boolean> {
    const liked: boolean = await this.repository.checkIfUserLikedPost(
      userId,
      postId,
    );

    return liked;
  }

  async getLikes(id: string): Promise<User[]> {
    const users: User[] = await this.repository.getLikes(id);

    return users;
  }
}
